hpx/lcos/channel.hpp

```diff
@@ -0,0 +1,397 @@
+                        )
+                    ) ||
+                (!data_.second && rhs.channel_ == nullptr) ||
+                (channel_ == nullptr && !rhs.data_.second);
+        }
+
+        void increment()
+        {
+            if (channel_)
+                data_ = get_checked();
+        }
+
+        typename base_type::reference dereference() const
+        {
+            HPX_ASSERT(data_.second);
+            return data_.first;
+        }
+
+    private:
+        Channel const* channel_;
+        std::pair<T, bool> data_;
+    };
+
+    template <typename Channel>
+    class channel_iterator<void, Channel>
+      : public hpx::util::iterator_facade<
+            channel_iterator<void, Channel>, util::unused_type const,
+            std::input_iterator_tag>
+    {
+        typedef hpx::util::iterator_facade<
+                channel_iterator<void, Channel>, util::unused_type const,
+                std::input_iterator_tag
+            > base_type;
+
+    public:
+        channel_iterator()
+          : channel_(nullptr), data_(false)
+        {}
+
+        channel_iterator(Channel const& c)
+          : channel_(&c), data_(get_checked())
+        {}
+
+    private:
+        bool get_checked()
+        {
+            hpx::future<void> f = channel_->get_async();
+            f.wait();
+            return !f.has_exception();
+        }
+
+        friend class hpx::util::iterator_core_access;
+
+        bool equal(channel_iterator const& rhs) const
+        {
+            return (data_ == rhs.data_ &&
+                        (channel_ == rhs.channel_ ||
+                            (channel_ != nullptr && rhs.channel_ != nullptr &&
+                             channel_->get_id() == rhs.channel_->get_id())
+                        )
+                    ) ||
+                (!data_ && rhs.channel_ == nullptr) ||
+                (channel_ == nullptr && !rhs.data_);
+        }
+
+        void increment()
+        {
+            if (channel_)
+                data_ = get_checked();
+        }
+
+        typename base_type::reference dereference() const
+        {
+            HPX_ASSERT(data_);
+            return util::unused;
+        }
+
+    private:
+        Channel const* channel_;
+        bool data_;
+    };
+
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename T>
+    class channel
+      : public components::client_base<channel<T>, lcos::server::channel<T> >
+    {
+        typedef components::client_base<
+                channel<T>, lcos::server::channel<T>
+            > base_type;
+
+        HPX_CONSTEXPR_OR_CONST
+        static std::size_t default_generation = std::size_t(-1);
+
+    public:
+        channel()
+        {}
+
+        // create a new instance of a channel component
+        channel(naming::id_type const& loc)
+          : base_type(hpx::new_<lcos::server::channel<T> >(loc))
+        {}
+        channel(lcos::future<naming::id_type> f)
```

Comment:
On Montag, 25. Juli 2016 11:13:20 CEST Hartmut Kaiser wrote:
> > +        typedef components::client_base<
> > +                channel<T>, lcos::server::channel<T>
> > +            > base_type;
> > +
> > +        HPX_CONSTEXPR_OR_CONST
> > +        static std::size_t default_generation = std::size_t(-1);
> > +
> > +    public:
> > +        channel()
> > +        {}
> > +
> > +        // create a new instance of a channel component
> > +        channel(naming::id_type const& loc)
> > +          : base_type(hpx::new_<lcos::server::channel<T> >(loc))
> > +        {}
> > +        channel(lcos::future<naming::id_type> f)
> 
> This is necessary for the following code to work (currently):
> 
>     hpx::lcos::channel c;
>     c = hpx:new_<hpx::lcos::server::channel>(hpx::find_here());
> 
> Thus it's enforced by the current creation APIs we have. If we changed those
> rules this public constructor could be avoided.

Well, this form isn't really needed as the first constructor does exactly that, 
no? I wouldn't think that our current creation APIs enforce it. It can be 
completely hidden, as your first ctor has shown.